src/v2/token.rs

```diff
@@ -0,0 +1,222 @@
+use std::{collections::HashMap, sync::Arc, time::SystemTime};
+
+use cache_control::CacheControl;
+use josekit::{jwk::JwkSet, jws::RS256, jwt, jwt::JwtPayload};
+use reqwest::{header, Client, Response};
+use serde::Deserialize;
+use time::OffsetDateTime;
+use tokio::sync::RwLock;
+use url::Url;
+
+/// Zitadel client to verify a token's validity
+#[derive(Debug, Clone)]
+pub struct ZitadelJWTVerifier {
+	/// Zitadel domain
+	domain: Url,
+	/// Client for performing the requests
+	client: Client,
+	/// Key set from Zitadel
+	jwks: Arc<RwLock<JwkSet>>,
+	/// Time when the jwks is no longer valid
+	expires_at: OffsetDateTime,
+}
+
+impl ZitadelJWTVerifier {
+	/// Creates a new verifier to verify with a specif server
+	#[must_use]
+	pub fn new(url: Url) -> Self {
+		let jwks = JwkSet::new();
+		Self {
+			domain: url,
+			client: Client::new(),
+			jwks: Arc::new(RwLock::new(jwks)),
+			expires_at: OffsetDateTime::now_utc(),
+		}
+	}
+
+	/// Verifies if a token is valid and returns the token payload
+	/// The performed verifications are:
+	///     - Token signature
+	///     - Token not expired
+	///     - Token not used before 'not before'
+	///     - Token issuer is the expected server
+	pub async fn verify(&mut self, token: String) -> Result<JwtPayload, TokenValidationError> {
+		let header = jwt::decode_header(&token)?;
+		let kid = header
+			.claim("kid")
+			.ok_or(TokenValidationError::BadToken("No kid"))?
+			.as_str()
+			.ok_or(TokenValidationError::BadToken("kid is not a string"))?;
+
+		let mut jwk = self.jwks.read().await.get(kid).first().map(|&jwk| jwk.clone());
+		if self.expires_at < OffsetDateTime::now_utc() || jwk.is_none() {
+			let (new_jwks, expires_at) = self.get_jwks().await?;
+			let mut jwks = self.jwks.write().await;
+			*jwks = new_jwks;
+			jwk = jwks.get(kid).first().map(|&jwk| jwk.clone());
+			self.expires_at = expires_at;
+		}
+
+		let Some(jwk) = jwk else {
+			return Err(TokenValidationError::KidNotFoundError);
+		};
+
+		let verifier =
+			RS256.verifier_from_jwk(&jwk).map_err(TokenValidationError::TokenDecodeError)?;
+		let (payload, _) = jwt::decode_with_verifier(token, &verifier)
+			.map_err(TokenValidationError::TokenDecodeError)?;
+
+		// Url always comes with an '/' at the end. We need to remove it before for
+		// checking
+		if !payload.issuer().is_some_and(|issuer| {
+			issuer == self.domain.as_str().strip_suffix("/").unwrap_or(self.domain.as_str())
+		}) {
+			return Err(TokenValidationError::TokenIssuerError(
+				payload.issuer().unwrap_or_default().to_owned(),
+			));
+		}
+
+		let now = OffsetDateTime::now_utc();
+		if !payload.expires_at().is_some_and(|t| t > now) {
+			return Err(TokenValidationError::TokenExpiredError);
+		}
```

Comment:
```suggestion
		payload.expires_at().is_some_and(|t| t > now)
			.then_some(()).ok_or(TokenExpiredError)?;
```

Reply:
Maybe also to distinguish errors from "missing required claim" errors:
```suggestion
		(payload.expires_at().ok_or(MissingClaim("exp"))? > now)
			.then_some(()).ok_or(TokenExpiredError)?;
```